spring-integration-core/src/main/java/org/springframework/integration/store/SimpleMessageStore.java

```diff
@@ -255,7 +292,9 @@ public MessageGroup removeMessageFromGroup(Object groupId, Message<?> messageToR
 				SimpleMessageGroup group = this.groupIdToMessageGroup.get(groupId);
 				Assert.notNull(group, "MessageGroup for groupId '" + groupId + "' " +
 						"can not be located while attempting to remove Message from the MessageGroup");
-				group.remove(messageToRemove);
+				if (group.remove(messageToRemove)) {
+	                groupUpperBound.release();
```

Comment:
The problem with this variable that it is **shared** between all groups, but according to the JavaDocs above it is:
```
* @param groupCapacity The capacity of each group.
```
So, we should use `UpperBound` with that size for the particular group.
That's why I've introduced a new variable to track that to fix the INT-3523.
Please, take a look to the https://github.com/artembilan/spring-integration/commit/60e13094ab952a95e1da4483cc5512c61a78a40e and let me know you position.